subprojects/core/src/main/java/org/gradle/internal/classpath/types/DefaultInstrumentingTypeRegistryFactory.java

```diff
@@ -0,0 +1,75 @@
+import org.gradle.cache.PersistentCache;
+import org.gradle.internal.classpath.ClasspathFileTransformer;
+import org.gradle.internal.classpath.ClasspathWalker;
+import org.gradle.internal.classpath.DefaultCachedClasspathTransformer.Convert;
+import org.gradle.internal.classpath.DefaultCachedClasspathTransformer.ParallelTransformExecutor;
+import org.gradle.internal.classpath.InstrumentingClasspathFileTransformer;
+import org.gradle.internal.vfs.FileSystemAccess;
+
+import java.io.File;
+import java.net.URL;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class DefaultInstrumentingTypeRegistryFactory implements InstrumentingTypeRegistryFactory {
+
+    private final InstrumentingTypeRegistry gradleCoreInstrumentingRegistry;
+    private final InstrumentingDirectSuperTypesCollector directSuperTypesCollector;
+
+    public DefaultInstrumentingTypeRegistryFactory(PersistentCache cache, ParallelTransformExecutor parallelTransformExecutor, ClasspathWalker classpathWalker, FileSystemAccess fileSystemAccess) {
+        this(new GradleCoreInstrumentingTypeRegistry(), cache, parallelTransformExecutor, classpathWalker, fileSystemAccess);
+    }
+
+    @VisibleForTesting
+    DefaultInstrumentingTypeRegistryFactory(InstrumentingTypeRegistry gradleCoreInstrumentingRegistry, PersistentCache cache, ParallelTransformExecutor parallelTransformExecutor, ClasspathWalker classpathWalker, FileSystemAccess fileSystemAccess) {
+        this.gradleCoreInstrumentingRegistry = gradleCoreInstrumentingRegistry;
+        this.directSuperTypesCollector = new DefaultInstrumentingDirectSuperTypesCollector(cache, parallelTransformExecutor, classpathWalker, fileSystemAccess);
+    }
+
+    @Override
+    public InstrumentingTypeRegistry createFor(Collection<URL> urls, ClasspathFileTransformer transformer) {
+        if (shouldReturnEmptyRegistry(transformer)) {
+            return InstrumentingTypeRegistry.empty();
+        }
+        List<File> files = urls.stream()
+            .filter(url -> url.getProtocol().equals("file"))
+            .map(Convert::urlToFile)
+            .collect(Collectors.toList());
+        return createFor(files, transformer);
+    }
+
+    @Override
+    public InstrumentingTypeRegistry createFor(List<File> files, ClasspathFileTransformer transformer) {
+        if (shouldReturnEmptyRegistry(transformer)) {
+            return InstrumentingTypeRegistry.empty();
+        }
+        Map<String, Set<String>> directSuperTypes = directSuperTypesCollector.visit(files, transformer.getFileHasher());
+        return new ExternalPluginsInstrumentingTypeRegistry(directSuperTypes, gradleCoreInstrumentingRegistry);
+    }
+
+    private boolean shouldReturnEmptyRegistry(ClasspathFileTransformer transformer) {
+        return gradleCoreInstrumentingRegistry.isEmpty() || !(transformer instanceof InstrumentingClasspathFileTransformer);
```

Comment:
Is there an assumption in the check for `gradleCoreInstrumentingRegistry.isEmpty()` that there will not be interceptor declaration with inheritance enabled on third-party receiver types? What if there are, but this registry is empty?

Reply:
Good catch! Yeah, I guess now there is such assumption. We can either forbid it in annotation processor or we allow it, and we generate some "inherited-instrumentation-exists.txt" file, that we use as a flag.

Do you plan to use that feature for third-party receiver types?

Reply:
We now forbid any non-gradle types to use `@InterceptInherited` annotation